pkg/provisioner/crdprovisioner.go

```diff
@@ -658,20 +658,29 @@ func (c *CrdProvisioner) detachVolume(ctx context.Context, azVAClient v1beta1.Az
 		return err
 	}
 
-	if azVolumeAttachment.Status.Detail != nil {
+	updateFunc := func(obj interface{}) error {
+		updateInstance := obj.(*diskv1beta1.AzVolumeAttachment)
+		updateInstance.Annotations[consts.VolumeDetachRequestAnnotation] = "crdProvisioner"
+		return nil
+	}
+	switch azVolumeAttachment.Status.State {
+	case diskv1beta1.Detaching:
 		// if detachment is pending, return to prevent duplicate request
-		if azVolumeAttachment.Status.State == diskv1beta1.Detaching {
-			return status.Errorf(codes.Aborted, "detachment still in process for volume (%s) and node (%s)", volumeName, nodeName)
+		return status.Errorf(codes.Aborted, "detachment still in process for volume (%s) and node (%s)", volumeName, nodeName)
+	case diskv1beta1.Attaching:
+		// if attachment is still happening in the background async, wait until attachment is complete
+		klog.Infof("Attachment for volume (%s) to node (%s) is in process... Waiting for the attachment to complete.", volumeName, nodeName)
+		if azVolumeAttachment, err = c.WaitForAttach(ctx, volumeID, nodeName); err != nil {
+			return err
 		}
-
-		klog.Infof("Requesting AzVolumeAttachment (%s) deletion", attachmentName)
-
-		updateFunc := func(obj interface{}) error {
+	case diskv1beta1.DetachmentFailed:
+		// if detachment failed, reset error and state to retrigger operation
+		updateFunc = func(obj interface{}) error {
+			_ = updateFunc(obj)
```

Comment:
Won't this end up being a recursive call since the closure captures `updateFunc` by reference and not by value?

Reply:
Oh, you are right. Good catch :)